core/shared/platform/common/posix/posix_thread.c

```diff
@@ -47,6 +54,164 @@ os_thread_wrapper(void *arg)
+
+    /* Get latest stack context (first in the list) for current thread */
+    bh_list *context_list = bh_hash_map_find(stack_context_handler.contexts,
+                                             (void *)(uintptr_t)self);
+    assert(context_list);
+    sigjmp_buf *context = bh_list_first_elem(context_list);
+    assert(context);
+
+    siglongjmp(*context, 1);
+}
+
+sigjmp_buf *
+os_create_stack_context(korp_tid handle)
+{
+    return BH_MALLOC(sizeof(sigjmp_buf));
+}
+
+static void
+free_stack_context_list(void *stack_context_list)
+{
+    bh_list *context_list = (bh_list *)stack_context_list;
+
+    sigjmp_buf *context = bh_list_first_elem(context_list);
+    while (context) {
+        sigjmp_buf *next = bh_list_elem_next(context);
+        BH_FREE(context);
+        context = next;
+    };
+
+    BH_FREE(context_list);
+}
+
+static inline void
+stack_ctx_block_sig_and_lock()
+{
+    sigemptyset(&newmask);
+    sigaddset(&newmask, STACK_CTX_SIG);
+    pthread_sigmask(SIG_BLOCK, &newmask, &oldmask);
+
+    os_mutex_lock(&stack_context_handler.lock);
+}
+
+static inline void
+stack_ctx_unlock_and_unblock_sig()
+{
+    os_mutex_unlock(&stack_context_handler.lock);
+    pthread_sigmask(SIG_SETMASK, &oldmask, NULL);
+}
+
+bool
+os_stack_contexts_init()
+{
+    struct sigaction act;
+    memset(&act, 0, sizeof(act));
+    act.sa_handler = stack_context_sig_handler;
+    sigfillset(&act.sa_mask);
+    if (sigaction(STACK_CTX_SIG, &act, NULL) < 0) {
+        LOG_ERROR("failed to set signal handler");
+        return false;
+    }
+
+    if (os_mutex_init(&stack_context_handler.lock) != 0) {
+        LOG_ERROR("failed to init context mutex");
+        return false;
+    }
+
+    stack_context_handler.contexts = bh_hash_map_create(
+        32, false, (HashFunc)thread_handle_hash,
+        (KeyEqualFunc)thread_handle_equal, NULL, free_stack_context_list);
+    if (stack_context_handler.contexts == NULL) {
+        LOG_ERROR("failed to init stack contexts");
+        return false;
+    }
+
+    return true;
+}
+
+void
+os_stack_contexts_destroy()
+{
+    if (!bh_hash_map_destroy(stack_context_handler.contexts))
+        LOG_ERROR("failed to destroy stack context map");
+    stack_context_handler.contexts = NULL;
+
+    if (os_mutex_destroy(&stack_context_handler.lock) != BHT_OK)
+        LOG_ERROR("failed to destroy stack contexts");
+}
+
+void
+os_remove_stack_context(korp_tid handle, sigjmp_buf *context)
+{
+    korp_tid self = pthread_self();
+
+    stack_ctx_block_sig_and_lock();
+
+    bh_list *context_list = bh_hash_map_find(stack_context_handler.contexts,
+                                             (void *)(uintptr_t)self);
+    assert(context_list);
+    int ret = bh_list_remove(context_list, context);
+    assert(ret == BH_LIST_SUCCESS);
+    BH_FREE(context);
+
+    stack_ctx_unlock_and_unblock_sig();
+}
+
+bool
+os_is_returning_from_signal(korp_tid handle, sigjmp_buf *context)
+{
+    if (sigsetjmp(*context, 1))
+        return true;
```

Comment:
this looks unsafe to me.
when we return here with a longjmp, the call frame of os_is_returning_from_signal is not valid anymore.

to make this work, os_is_returning_from_signal should always be inlined.
maybe it's more straightforward to use a macro instead.

Reply:
thanks, that might explain some problems that I was seeing when trying to use it with AOT. In any case, I realized that I missed this  PR #1309 that is basically doing something very similar to what I want to do. The approach is better because we don't need to keep a map <thread_id, stack contexts> if we use tls. I'll try to apply a similar approach.

Reply:
@eloparco Yes, in the memory access boundary check with hardware trap feature, both interpreter and AOT use the setjmp/longjmp mechanism in posix like platforms. The local variable jmpbuf is pushed into exec_env's jmpbuf stack before setjmp, runtime doesn't allocate memory for it.
I think that the principle of this PR is similar to the hw bound check feature: firstly register a signal handler, then setjmp before call, and when running wasm function, if there is a signal raised, the signal handler is trigger and the handler longjmp into the place of setjmp.

I think there may be two ways to implement this PR: one is to extend the hw bound check feature, e.g. reuse the current signal handler, the current setjmp and longjmp process, but add another macro to control them respectively (for example, OS_ENABLE_XXX and OS_ENABLE_HW_BOUND_CHECK). The other is to register new signal handler, add new setjmp/longjmp process, but I suggest not to add setjmp in wasm_interp_class.c, had better add in wasm_runtime.c, before calling wasm function:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_runtime.c#L2163
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_runtime.c#LL2459C3-L2459C3

And note that setjmp is only supported by posix platforms, there should be macro to control the feature, and had better use os_setjmp/os_longjmp.

Reply:
And could you submit this PR to main branch since the main branch also has the issue?

Reply:
> https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_runtime.c#L2163

yes that works, for aot I'm not sure what to wrap with `os_setjmp` since there are 2 calls, I need to check https://github.com/bytecodealliance/wasm-micro-runtime/blob/fe3347d5d2d92c3a78d77c55877692bc07ac8888/core/iwasm/aot/aot_runtime.c#L1411
https://github.com/bytecodealliance/wasm-micro-runtime/blob/fe3347d5d2d92c3a78d77c55877692bc07ac8888/core/iwasm/aot/aot_runtime.c#L1470

Reply:
> And could you submit this PR to main branch since the main branch also has the issue?

I'll try to have it working here first since I'm testing it with wasi threads, then I'll create a PR in main to merge it there

Reply:
I'm closing this PR in favour of #1930, where I rely on what's already implemented for hardware bound checks.